File: languages/dotnet/runtime/Revaly.Sdk/Testing/RecordedRequest.cs
namespace Revaly.Sdk.Testing;

/// <summary>One request observed by the mock transport, for merchant test assertions.</summary>
/// <param name="Method">The HTTP method.</param>
/// <param name="Path">The request path.</param>
/// <param name="Headers">A snapshot of the request headers.</param>
/// <param name="Body">The request body, when present.</param>
public sealed record RecordedRequest(
    string Method,
    string Path,
    IReadOnlyDictionary<string, string> Headers,
    string? Body);

File: languages/php/runtime/src/Testing/RecordedRequest.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Testing;

use Psr\Http\Message\RequestInterface;

/** One request observed by {@see RapMockTransport}, for merchant test assertions. */
final class RecordedRequest
{
    /**
     * @param array<string, array<string>> $headers
     */
    public function __construct(
        public readonly string $method,
        public readonly string $uri,
        public readonly string $path,
        public readonly array $headers,
        public readonly string $body,
    ) {
    }

    public static function from(RequestInterface $request): self
    {
        return new self(
            $request->getMethod(),
            (string) $request->getUri(),
            $request->getUri()->getPath(),
            $request->getHeaders(),
            (string) $request->getBody(),
        );
    }

    /** The first value of the named header (case-insensitive), or null. */
    public function header(string $name): ?string
    {
        foreach ($this->headers as $headerName => $values) {
            if (strcasecmp($headerName, $name) === 0) {
                return $values[0] ?? null;
            }
        }

        return null;
    }
}
